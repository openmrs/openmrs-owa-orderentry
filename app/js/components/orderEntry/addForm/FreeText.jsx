import React from 'react';
import PropTypes from 'prop-types';

const FreeText = ({
  fields,
  allConfigurations,
  handleChange,
  handleSubmit,
}) => {
  const { completeInstructions, dispensingQuantity, dispensingUnit } = fields;
  return (
    <div>
      <form>
        <div>
          <div>
            <label name="instructions">Complete Instructions</label>
            <textarea
              id="instructions"
              name="completeInstructions"
              value={completeInstructions}
              onChange={handleChange}
              required
            />
          </div>
          <div>
            <div>
              <label>Dispense:</label>
            </div>
            <div>
              <label name="Quantity">Dose</label>
              <input
                type="number"
                id="Quantity"
                name="dispensingQuantity"
                min="0"
                value={dispensingQuantity}
                onChange={handleChange}
                required
              />
            </div>
            <div>
              <label name="dispenseUnits">Units</label>
              <input
                type="search"
                id="dispenseUnits"
                name="dispensingUnit"
                list="units"
                size="10"
                value={dispensingUnit}
                onChange={handleChange}
                required />
              <datalist id="units" >
                {
                  allConfigurations &&
                  (
                    allConfigurations.drugDispensingUnits &&
                  allConfigurations.drugDispensingUnits.map(unit =>
                    (
                      <option key={unit.uuid} value={unit.display} />
                    ))
                  )
                }
              </datalist>
            </div>
          </div>
          <div>
            <input type="button" className="button cancel" value="Cancel" />
            <input type="button" onClick={handleSubmit} className="button confirm right" value="Save" />
          </div>
        </div>
      </form>
    </div>
  );
};

FreeText.propTypes = {
  fields: PropTypes.object.isRequired,
  allConfigurations: PropTypes.shape({
    drugDispensingUnits: PropTypes.array,
  }),
  handleChange: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
};

FreeText.defaultProps = {
  allConfigurations: {},
};

export default FreeText;
