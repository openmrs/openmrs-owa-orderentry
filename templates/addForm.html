<!DOCTYPE html>
<html lang="en">

<head>
  <!-- Required meta tags -->
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <title>Order Entry</title>

  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
</head>

<body>
  <div class="container-fluid">
    <div class="row">
      <div class="col-lg-8 col-lg-offset-2">
        <div class="card">
          <div class="card-body">
            <ul class="nav nav-pills mb-3" id="pills-tab" role="tablist">
              <li class="nav-item">
                <a class="nav-link" id="standard-tab" data-toggle="pill" href="#standard" role="tab" aria-controls="standard" aria-selected="false">Standard Dosing</a>
              </li>
              <li class="nav-item">
                <a class="nav-link" id="free-tab" data-toggle="pill" href="#free" role="tab" aria-controls="free" aria-selected="false">Free Text</a>
              </li>
              </li>
            </ul>
            <br>

            <div class="tab-content" id="pills-tabContent">
              <div class="tab-pane fade" id="free" role="tabpanel" aria-labelledby="free-tab">
                <form>
                  <div class="row">
                    <div class="col-sm-11">
                      <div class="form-group">
                        <label for="instructions">Completete Instructions:</label>
                        <textarea class="form-control" rows="5" id="instructions"></textarea>
                      </div>
                    </div>
                  </div>
                </form>

                <div class="row">
                  <div class="col-sm-3 text-left">
                    <button type="button" class="btn btn-danger">Cancel</button>
                  </div>
                  <div class="col-sm-3 col-sm-offset-5 text-right">
                    <button type="button" class="btn btn-success">Save</button>
                  </div>
                </div>
              </div>

              <div class="tab-pane fade" id="standard" role="tabpanel" aria-labelledby="standard-tab">
                <form>
                  <div class="row">
                    <div class="col-sm-1">
                      <div class="form-group">
                        <label for="dose">Dose</label>
                        <input type="text" class="form-control" id="dose">
                      </div>
                    </div>
                    <div class="col-sm-2">
                      <div class="form-group">
                        <label for="unit">Unit</label>
                        <select class="form-control" id="unit">
                          <option>mg</option>
                          <option>ml</option>
                        </select>
                      </div>
                    </div>
                    <div class="col-sm-3">
                      <div class="form-group">
                        <label for="frequency">Frequency</label>
                        <select class="form-control" id="frequency">
                          <option>Once a day</option>
                          <option>Twice a day</option>
                        </select>
                      </div>
                    </div>
                    <div class="col-sm-2">
                      <div class="form-group">
                        <label for="route">Route</label>
                        <select class="form-control" id="route">
                          <option>Oral</option>
                          <option>Inhalation</option>
                        </select>
                      </div>
                    </div>
                    <div class="col-sm-1">
                      <div class="form-group">
                        <label for="duration">Duration</label>
                        <input type="text" class="form-control" id="duration">
                      </div>
                    </div>
                    <div class="col-sm-2">
                      <div class="form-group">
                        <label for="days"></label>
                        <select class="form-control" id="days">
                          <option>Days</option>
                          <option>Weeks</option>
                        </select>
                      </div>
                    </div>
                  </div>

                  <div class="row">
                    <div class="col-sm-5">
                      <div class="form-group">
                        <label for="needed">Needed for</label>
                        <input type="text" class="form-control" id="needed">
                      </div>
                    </div>
                    <div class="col-sm-2 text-right">
                      <medium class="form-text "> Dispense: </medium>
                    </div>
                    <div class="col-sm-1">
                      <div class="form-group">
                        <label for="dose">Dose</label>
                        <input type="text" class="form-control" id="dose">
                      </div>
                    </div>
                    <div class="col-sm-2">
                      <div class="form-group">
                        <label for="unit">Unit</label>
                        <select class="form-control" id="unit">
                          <option>mg</option>
                          <option>ml</option>
                        </select>
                      </div>
                    </div>
                  </div>

                  <div class="row">
                    <div class="col-sm-2">
                      <medium class="form-text"> Start Date: </medium>
                    </div>
                  </div>

                  <div class="row">
                    <div class="col-sm-2">
                      <div class="form-check text-right">
                        <input class="form-check-input" type="radio" name="today" id="today" value="option1" checked>
                        <label class="form-check-label text-right" for="today">
                          Today
                        </label>
                      </div>
                    </div>
                    <div class="col-sm-2">
                      <div class="form-check text-right">
                        <input class="form-check-input" type="radio" name="otherDate" id="otherDate" value="option2">
                        <label class="form-check-label text-right" for="otherDate">
                          Other Date
                        </label>
                      </div>
                    </div>
                    <div class="col-sm-2">
                      <div class="form-group">
                        <div class='input-group date' id='datetimepicker1'>
                          <input type='text' class="form-control" />
                          <span class="input-group-addon">
                            <span class="glyphicon glyphicon-calendar"></span>
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>

                  <div class="row">
                    <div class="col-sm-11">
                      <div class="form-group">
                        <label for="notes">Notes:</label>
                        <textarea class="form-control" rows="4" id="notes"></textarea>
                      </div>
                    </div>
                  </div>
                </form>

                <div class="row">
                  <div class="col-sm-3 text-left">
                    <button type="button" class="btn btn-danger">Cancel</button>
                  </div>
                  <div class="col-sm-3 col-sm-offset-5 text-right">
                    <button type="button" class="btn btn-success">Save</button>
                  </div>
                </div>
              </div>
            </div>

          </div>
        </div>
      </div>
    </div>
  </div>
</body>

</html>
