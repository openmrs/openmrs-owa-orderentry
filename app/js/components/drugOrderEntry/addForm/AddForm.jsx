import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import FreeText from './FreeText';
import StandardDose from './StandardDose';
import DosageTabs from '../../tabs/DosageTabs';
import DosageTab from '../../tabs/DosageTab';
import getOrderEntryConfigurations from '../../../actions/orderEntryActions';
import { addDraftOrder } from '../../../actions/draftTableAction';
import { selectDrugSuccess } from '../../../actions/drug';
import { setOrderAction, setSelectedOrder } from '../../../actions/orderAction';

export class AddForm extends React.Component {
  state = {
    action: 'NEW',
    fields: {
      dose: '',
      dosingUnit: '',
      frequency: '',
      route: '',
      duration: '',
      durationUnit: '',
      dispensingQuantity: '',
      dispensingUnit: '',
      reason: '',
      drugInstructions: '',
    },
    previousOrder: null,
    fieldErrors: {
    },
    draftOrder: {},
    orders: [],
    orderNumber: 0,
    formType: 'Standard Dosage',
    dosingType: '',
    activeTabIndex: 0,
    options: {
      dosingUnit: [],
      frequency: [],
      route: [],
      durationUnit: [],
      dispensingUnit: [],
    },
  };

  componentDidMount() {
    this.props.getOrderEntryConfigurations();
  }

  componentDidUpdate(prevProps) {
    return (
      Object.keys(this.props.draftOrder).length ||
      Object.keys(this.props.editOrder).length
    ) && this.populateEditOrderForm();
  }

  handleFormTabs = (tabIndex) => {
    this.setState({
      activeTabIndex: tabIndex,
      fieldErrors: {},
    });
  }

  handleFormType = (formType) => {
    this.setState({ formType: formType.trim() });
  }

  handleSubmitDrugForm = () => {
    const {
      dose,
      dosingUnit,
      frequency,
      route,
      duration,
      durationUnit,
      dispensingQuantity,
      dispensingUnit,
      reason,
      drugInstructions,
    } = this.state.fields;

    this.setState({
      draftOrder: {
        drugName: this.props.drugName,
        action: this.state.action,
        dose,
        dosingUnit,
        frequency,
        route,
        duration,
        durationUnit,
        dispensingUnit,
        dispensingQuantity,
        reason,
        previousOrder: this.state.previousOrder,
        drugInstructions,
        careSetting: this.props.careSetting.uuid,
        drug: this.props.drugUuid,
        orderer: this.props.session.currentProvider.uuid,
        dosingType: this.state.formType === 'Standard Dosage' ?
          'org.openmrs.SimpleDosingInstructions' :
          'org.openmrs.FreeTextDosingInstructions',
        type: "drugorder",
        orderNumber: (this.state.action === 'NEW') ? this.props.draftOrders.length :
          this.props.orderNumber,
      },
    }, () => {
      this.props.addDraftOrder(this.state.draftOrder);
      if (this.state.draftOrder.action === 'NEW') {
        this.props.setOrderAction('DRAFT', this.state.orderNumber);
      } else {
        this.props.setOrderAction('DRAFT_EDIT', this.props.orderNumber);
      }
    });
    this.props.selectDrugSuccess('');
    this.clearDrugForms();
    this.props.clearSearchField();
    if (this.props.activity === 'DRAFT_ORDER_EDIT') {
      this.props.setSelectedOrder({
        currentOrderType: this.props.currentOrderType,
        selectedOrder: null,
        activity: null,
      });
    }
  }


  /**
   * Validation of datalist tag values using onblur event handler
   */
  handleValidation = (event) => {
    const {
      drugDosingUnits, orderFrequencies, drugRoutes, durationUnits, drugDispensingUnits,
    } = this.props.allConfigurations;
    const { name, value } = event.target;
    let item = false;
    const options = [];
    switch (name) {
      case 'dosingUnit':
        item = drugDosingUnits.filter(unit => unit.display === value).length > 0;
        break;
      case 'frequency':
        item = orderFrequencies.filter(frequency => frequency.display === value).length > 0;
        break;
      case 'route':
        item = drugRoutes.filter(route => route.display === value).length > 0;
        break;
      case 'durationUnit':
        item = durationUnits.filter(unit => unit.display === value).length > 0;
        break;
      case 'dispensingUnit':
        item = drugDispensingUnits.filter(unit => unit.display === value).length > 0;
        break;
      default:
        item = value.length > 0;
    }

    { this.setState({
      ...this.state,
      fields: {
        ...this.state.fields, [name]: item ? value : '',
      },
      fieldErrors: {
        ...this.state.fieldErrors, [name]: !item,
      },
      options: {
        ...this.state.options, [name]: options,
      },
    }); }
  }

  handleChange = (event) => {
    const { name, value } = event.target;
    const {
      drugDosingUnits, orderFrequencies, drugRoutes, durationUnits, drugDispensingUnits,
    } = this.props.allConfigurations;
    let options;
    if (value) {
      switch (name) {
        case 'dosingUnit':
          options = drugDosingUnits.filter(unit => (
            unit.display.toLowerCase().indexOf(value.toLowerCase()) === 0));
          break;
        case 'frequency':
          options = orderFrequencies.filter(frequency => (
            frequency.display.toLowerCase().indexOf(value.toLowerCase()) === 0));
          break;
        case 'route':
          options = drugRoutes.filter(route => (
            route.display.toLowerCase().indexOf(value.toLowerCase()) === 0));
          break;
        case 'dispensingUnit':
          options = drugDispensingUnits.filter(unit => (
            unit.display.toLowerCase().indexOf(value.toLowerCase()) === 0));
          break;
        case 'durationUnit':
          options = durationUnits.filter(unit => (
            unit.display.toLowerCase().indexOf(value.toLowerCase()) === 0));
          break;
        default:
          // does nothing
      }
    }

    this.setState({
      ...this.state,
      fields: { ...this.state.fields, [name]: value },
      options: { ...this.state.options, [name]: options },
    });
  }

  handleCancel = () => {
    this.props.selectDrugSuccess('');
    this.clearDrugForms();
    this.props.clearSearchField();
    this.props.setOrderAction('DISCARD_ONE', this.props.orderNumber);
    this.props.setSelectedOrder({
      currentOrderType: this.props.currentOrderType,
      selectedOrder: null,
      activity: null,
    });
  }
  clearDrugForms = () => {
    this.setState({
      action: 'NEW',
      fields: {
        dose: '',
        dosingUnit: '',
        frequency: '',
        route: '',
        duration: '',
        durationUnit: '',
        dispensingQuantity: '',
        dispensingUnit: '',
        reason: '',
        drugInstructions: '',
      },
      previousOrder: null,
      fieldErrors: {
      },
      orderNumber: 0,
      formType: 'Standard Dosage',
      dosingType: '',
      activeTabIndex: 0,
      options: {
        dosingUnit: [],
        frequency: [],
        route: [],
        durationUnit: [],
        dispensingUnit: [],
      },
    });
  }

  activateSaveButton = () => {
    const {
      dose,
      dosingUnit,
      frequency,
      route,
      dispensingQuantity,
      dispensingUnit,
      duration,
      durationUnit,
      drugInstructions,
    } = this.state.fields;

    if (Object.values(this.state.fieldErrors).includes(true)) {
      return true;
    } else if (this.state.formType === 'Standard Dosage') {
      if ((duration && !durationUnit) ||
        !(
          dose &&
          dosingUnit &&
          frequency &&
          route &&
          dispensingQuantity &&
          dispensingUnit
        )) {
        return true;
      }
    } else if (this.state.formType === 'Free Text' &&
      !(drugInstructions && dispensingQuantity && dispensingUnit)) {
      return true;
    }
    return false;
  }

  populateEditOrderForm = () => {
    const { editOrder, draftOrder } = this.props;
    this.setState({
      activeTabIndex: (draftOrder.dosingType || editOrder.dosingType) === 'org.openmrs.SimpleDosingInstructions' ? 0 : 1,
      action: draftOrder.action || editOrder.action,
      previousOrder: editOrder.uuid || ((draftOrder.action === 'REVISE') && draftOrder.previousOrder) || null,
      formType: (draftOrder.dosingType || editOrder.dosingType) === 'org.openmrs.SimpleDosingInstructions' ? 'Standard Dosage' : 'Free Text',
      dosingType: draftOrder.dosingType || editOrder.dosingType,
      fields: {
        dose: editOrder.dose || draftOrder.dose || '',
        dosingUnit: (editOrder.doseUnits && editOrder.doseUnits.display) || draftOrder.dosingUnit || '',
        frequency: (editOrder.frequency && editOrder.frequency.display) || draftOrder.frequency || '',
        route: (editOrder.route && editOrder.route.display) || draftOrder.route || '',
        duration: editOrder.duration || draftOrder.duration || '',
        durationUnit: (editOrder.durationUnits && editOrder.durationUnits.display) || draftOrder.durationUnit || '',
        dispensingQuantity: editOrder.quantity || draftOrder.dispensingQuantity || '',
        dispensingUnit: (editOrder.quantityUnits && editOrder.quantityUnits.display) || draftOrder.dispensingUnit || '',
        reason: editOrder.asNeededCondition || draftOrder.reason || '',
        drugInstructions: editOrder.dosingInstructions || draftOrder.drugInstructions || '',
      },
    });
    this.props.removeOrder();
  }

  renderDrugOrderForms = () => (
    <div>
      <DosageTabs
        formType={this.handleFormType}
        activeTabIndex={this.state.activeTabIndex}
        handleFormTabs={this.handleFormTabs}
      >
        <DosageTab tabName="Standard Dosage &nbsp;" icon="icon-th-large">
          <StandardDose
            fields={this.state.fields}
            fieldErrors={this.state.fieldErrors}
            options={this.state.options}
            handleValidation={this.handleValidation}
            activateSaveButton={this.activateSaveButton}
            handleChange={this.handleChange}
            handleSubmit={this.handleSubmitDrugForm}
            handleCancel={this.handleCancel} />
          <br />
        </DosageTab>
        <DosageTab tabName="Free Text" icon="icon-edit">
          <FreeText
            fields={this.state.fields}
            fieldErrors={this.state.fieldErrors}
            options={this.state.options}
            handleValidation={this.handleValidation}
            activateSaveButton={this.activateSaveButton}
            handleChange={this.handleChange}
            handleCancel={this.handleCancel}
            handleSubmit={this.handleSubmitDrugForm}
          />
          <br />
        </DosageTab>
      </DosageTabs>
    </div>
  );

  render() {
    return (
      <div>
        { this.props.drugUuid && this.renderDrugOrderForms() }
      </div>
    );
  }
}

const mapStateToProps = ({
  orderEntryConfigurations,
  drugSearchReducer,
  draftReducer: { draftDrugOrders },
  openmrs: { session },
  orderSelectionReducer: { activity },
}) =>
  ({
    activity,
    drug: drugSearchReducer.selected,
    draftOrders: draftDrugOrders,
    allConfigurations: ((orderEntryConfigurations || {}).configurations || {}),
    session,
  });

AddForm.propTypes = {
  activity: PropTypes.string,
  clearSearchField: PropTypes.func.isRequired,
  currentOrderType: PropTypes.object.isRequired,
  selectDrugSuccess: PropTypes.func,
  getOrderEntryConfigurations: PropTypes.func,
  addDraftOrder: PropTypes.func.isRequired,
  setOrderAction: PropTypes.func.isRequired,
  orderNumber: PropTypes.number,
  draftOrders: PropTypes.arrayOf(PropTypes.any),
  allConfigurations: PropTypes.object.isRequired,
  drugName: PropTypes.string,
  careSetting: PropTypes.object.isRequired,
  drugUuid: PropTypes.string,
  removeOrder: PropTypes.func.isRequired,
  editOrder: PropTypes.shape({
    dose: PropTypes.number,
    doseUnits: PropTypes.shape({}),
    frequency: PropTypes.shape({}),
    route: PropTypes.shape({}),
    duration: PropTypes.number,
    durationUnits: PropTypes.shape({}),
    quantity: PropTypes.number,
    quantityUnits: PropTypes.shape({}),
    asNeededCondition: PropTypes.string,
    dosingInstructions: PropTypes.string,
    dosingType: PropTypes.string,
  }),
  draftOrder: PropTypes.shape({}),
  setSelectedOrder: PropTypes.func.isRequired,
  session: PropTypes.shape({
    currentProvider: PropTypes.shape({
      uuid: PropTypes.string,
    }),
  }).isRequired,
};

AddForm.defaultProps = {
  activity: '',
  selectDrugSuccess: {},
  getOrderEntryConfigurations: () => {},
  drugName: '',
  editOrder: {},
  draftOrder: {},
  draftOrders: [],
  drugUuid: '',
  orderNumber: '0',
};

export default connect(
  mapStateToProps,
  {
    getOrderEntryConfigurations,
    selectDrugSuccess,
    addDraftOrder,
    setOrderAction,
    setSelectedOrder,
  },
)(AddForm);
