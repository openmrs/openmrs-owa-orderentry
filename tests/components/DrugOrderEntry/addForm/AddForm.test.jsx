import React from 'react';

import { AddForm } from '../../../../app/js/components/drugOrderEntry/addForm/AddForm';

const { draftOrders, draftOrder, editOrder, formType, session } = mockData;

const props = {
  addDraftOrder: jest.fn(),
  clearDrugForms: jest.fn(),
  clearSearchField: jest.fn(),
  setOrderAction: jest.fn(),
  removeOrder: jest.fn(),
  selectDrugSuccess: jest.fn(),
  getOrderEntryConfigurations: jest.fn(),
  allConfigurations: {
    drugDosingUnits: [{display: 'grams'}],
    orderFrequencies: [{display: 'daily'}],
    drugRoutes: [{display: 'eye drops'}],
    drugDispensingUnits: [{display: 'kits'}],
    durationUnits:  [{display: 'months'}],
  },
  careSetting: {display: 'Inpatient'},
  drugName: 'Paracentamol',
  drugUuid: 'AJJJKW7378JHJ',
  draftOrder,
  draftOrders,
  editOrder,
  formType,
  session,
};

const outpatientprops = {
  ...props,
  careSetting: {display: 'Outpatient'},
}

let mountedComponent;
const getComponent = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(<AddForm {...props} />);
  }
  return mountedComponent;
};

describe('Test for adding a new drug order', () => {
    it('should render component', () => {
      const wrapper = shallow(<AddForm {...props} />  );
      expect(wrapper).toMatchSnapshot()
    });
    describe('inpatient orders', () =>{
      const wrapper = mount(<AddForm {...props} />  );
      it('Confirm button is originally disabled', () => {
        expect(wrapper.find('button.confirm').props().disabled).toBe(true);
      });
      it('should not have any errors messages initially', () => {
        expect(wrapper.find('span.field-error').length).toBe(0);
      });
      it('should not have options initially', () => {
        expect(wrapper.find('option')).toHaveLength(0);
      });
      it('should have options once a user inputs a unit that corresponds to configurations', () => {
        wrapper.find('[name="dosingUnit"]').simulate('change', {target: {name: 'dosingUnit', value: 'g'}});
        expect(wrapper.find('option')).toHaveLength(1);

      });
      describe('Activate and deactivate Confirm button', () => {
        beforeEach(() => {
          wrapper.find('[name="dose"]').simulate('change', {target: {name: 'dose', value: 8}});
          wrapper.find('[name="dosingUnit"]').simulate('change', {target: {name: 'dosingUnit', value: 'kilogram'}});
          wrapper.find('[name="route"]').simulate('change', {target: {name: 'route', value: 'oral'}});
          wrapper.find('[name="frequency"]').simulate('change', {target: {name: 'frequency', value: 'once'}});
        });
        it('should activate when the required fields are filled', () => {
          expect(wrapper.find('button.confirm').props().disabled).toBe(false);
        });
        it('should deactivate with duration without units', () => {
          wrapper.find('[name="duration"]').simulate('change', {target: {name: 'duration', value: 7}});
          expect(wrapper.find('button.confirm').props().disabled).toBe(true);
        });
        it('should activate with both duration and units', () => {
          wrapper.find('[name="duration"]').simulate('change', {target: {name: 'duration', value: 7}});
          wrapper.find('[name="durationUnit"]').simulate('change', {target: {name: 'durationUnit', value: 'weeks'}});
          expect(wrapper.find('button.confirm').props().disabled).toBe(false);
        });
        it('should deactivate with both duration and invalid units', () => {
          wrapper.find('[name="duration"]').simulate('change', {target: {name: 'duration', value: 7}});
          wrapper.find('[name="durationUnit"]').simulate('change', {target: {name: 'durationUnit', value: 'weeks'}});
          wrapper.find('[name="durationUnit"]').simulate('blur');
          expect(wrapper.find('button.confirm').props().disabled).toBe(true);
        });
      });
    }); 
   describe('Outpatient orders', () => {
      const wrapper = mount(<AddForm {...outpatientprops} />);
      describe('Activate and deactivate Confirm button under standard dosing form', () => {
        beforeEach(() => {
          wrapper.find('[name="dose"]').simulate('change', {target: {name: 'dose', value: 8}});
          wrapper.find('[name="dosingUnit"]').simulate('change', {target: {name: 'dosingUnit', value: 'kilogram'}});
          wrapper.find('[name="route"]').simulate('change', {target: {name: 'route', value: 'oral'}});
          wrapper.find('[name="frequency"]').simulate('change', {target: {name: 'frequency', value: 'once'}});
        });
        it('should be deactivated without both dispensing quantity and units', () => {
          expect(wrapper.find('button.confirm').props().disabled).toBe(true);
        });
        it('should be deactivated with dispensing quantity without units', () => {
          wrapper.find('[name="dispensingQuantity"]').simulate('change', {target: {name: 'dispensingQuantity', value: 12}});
          expect(wrapper.find('button.confirm').props().disabled).toBe(true);
        });
        it('should be activated with both dispensing Quantity and units', () => {
          wrapper.find('[name="dispensingQuantity"]').simulate('change', {target: {name: 'dispensingQuantity', value: 12}});
          wrapper.find('[name="dispensingUnit"]').simulate('change', {target: {name: 'dispensingUnit', value: 'kits'}});
          expect(wrapper.find('button.confirm').props().disabled).toBe(false);
        });
      });
      describe('Activate and deactivate Confirm button under free text form', () => {
        it('should activate when the required fields are filled', () => {
          wrapper.setState({ formType: 'Free Text' })
          expect(wrapper.state().formType).toEqual('Free Text');
          wrapper.find('[name="drugInstructions"]').simulate('change', {target: {name: 'drugInstructions', value: "3 tablets"}});
          wrapper.find('[name="dispensingQuantity"]').simulate('change', {target: {name: 'dispensingQuantity', value: 12}});
          wrapper.find('[name="dispensingUnit"]').simulate('change', {target: {name: 'dispensingUnit', value: 'kits'}});
          expect(wrapper.find('button.confirm').props().disabled).toBe(false);
        });
      });
      describe('Validation of fields', () => {
        beforeEach(() => {
          wrapper.find('[name="dispensingQuantity"]').simulate('change', {target: {name: 'dispensingQuantity', value: 12}});
          wrapper.find('[name="dispensingUnit"]').simulate('change', {target: {name: 'dispensingUnit', value: 'tins'}});
        });
        it ('should deactivate confirm button with invalid dose units', () => {
          wrapper.find('[name="dosingUnit"]').simulate('blur')
          expect(wrapper.find('button.confirm').props().disabled).toBe(true);
        });
        it ('should display error with invalid dosing units', () => {
          wrapper.find('[name="dosingUnit"]').simulate('blur')
          expect(wrapper.find("span.field-error").length).toBe(1);
        });
        it ('should activate with valid dose units', () => {
          wrapper.find('[name="dosingUnit"]').simulate('change', {target: {name: 'dosingUnit', value: 'grams'}});
          wrapper.find('[name="dosingUnit"]').simulate('blur');
          expect(wrapper.find('button.confirm').props().disabled).toBe(false);
        });
        it ('should deactivate with invalid frequency', () => {
          wrapper.find('[name="frequency"]').simulate('blur');
          expect(wrapper.find('button.confirm').props().disabled).toBe(true);
        });
        it ('should deactivate with invalid route', () => {
          wrapper.find('[name="route"]').simulate('blur');
          expect(wrapper.find('button.confirm').props().disabled).toBe(true);
        });
        it ('should deactivate with invalid dispensing unit', () => {
          wrapper.find('[name="dispensingUnit"]').simulate('blur');
          expect(wrapper.find('button.confirm').props().disabled).toBe(true);
        });
      });
  });
});

describe('Test AddForm state', () => {
  it('should update state', () => {
    mountedComponent = mount(<AddForm {...props} />);
    getComponent().setState({fields:  { dose: 11}})
    getComponent().find('[name="dose"]').simulate('change', {target: {name: 'dose', value: 9}});
    getComponent().find('.confirm').simulate('click');
    expect(getComponent().state().fields.dose).toEqual(9);
  });
});

describe('handleFormType() method', () => {
  it('should call handleFormType()', () => {
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'handleFormType');
    renderedComponent.handleFormType(formType);
    expect(renderedComponent.handleFormType.calledOnce).toEqual(true);
    expect(getComponent().state('formType')).toEqual("Free Text");
  });
});

describe('handleFormTabs() method', () => {
  it('should call handleFormTabs()', () => {
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'handleFormTabs');
    renderedComponent.handleFormTabs(1);
    expect(renderedComponent.handleFormTabs.calledOnce).toEqual(true);
    expect(getComponent().state('activeTabIndex')).toEqual(1);
  });
});

describe('handleCancel() method', () => {
  it('should call handleCancel()', () => {
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'handleCancel');
    renderedComponent.handleCancel();
    expect(renderedComponent.handleCancel.calledOnce).toEqual(true);
  });
});

describe('populateEditOrderForm() method', () => {
  it('should call populateEditActiveOrderForm()', () => {
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'populateEditOrderForm');
    renderedComponent.populateEditOrderForm();
    expect(renderedComponent.populateEditOrderForm.calledOnce).toEqual(true);
    expect(getComponent().state('activeTabIndex')).toEqual(1);
    expect(getComponent().state('fields')).toEqual({
      dispensingQuantity: "",
      dispensingUnit: "",
      dose: "",
      dosingUnit: "",
      drugInstructions: "",
      duration: "",
      durationUnit: "",
      frequency: "",
      reason: "",
      route: "",
    });
  });
});

describe('clearDrugForms() method', () => {
  it('should call clearDrugForms()', () => {
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'clearDrugForms');
    renderedComponent.clearDrugForms();
    expect(renderedComponent.clearDrugForms.calledOnce).toEqual(true);
    expect(getComponent().state('fields')).toEqual({
      dispensingQuantity: "",
      dispensingUnit: "",
      dose: "",
      dosingUnit: "",
      drugInstructions: "",
      duration: "",
      durationUnit: "",
      frequency: "",
      reason: "",
      route: "",
    });
  });
});

describe('handleSubmitDrugForm() method', () => {
  it('should call handleSubmitDrugForm()', () => {
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'handleSubmitDrugForm');
    renderedComponent.handleSubmitDrugForm();
    expect(renderedComponent.handleSubmitDrugForm.calledOnce).toEqual(true);
    expect(getComponent().state('draftOrder')).toEqual({
      action: "NEW",
      careSetting: undefined,
      dosingType: "org.openmrs.SimpleDosingInstructions",
      drug: "AJJJKW7378JHJ",
      drugName: "Paracentamol",
      orderNumber: 1,
      type: "drugorder",
      orderer: "",
      previousOrder: null,
      dispensingQuantity: "",
      dispensingUnit: "",
      dose: "",
      dosingUnit: "",
      drugInstructions: "",
      duration: "",
      durationUnit: "",
      frequency: "",
      reason: "",
      route: "",
    });
  });
});

describe('componentWillReceiveProps()', () => {
  it('does not call clearDrugForms when props is the same', () => {
      const renderedComponent = getComponent().instance();
      getComponent().setProps({ careSetting: { display: 'Inpatient' } })
      expect(renderedComponent.clearDrugForms.calledOnce).toEqual(false);

      getComponent().setProps({ careSetting: { display: 'Inpatient' } })
      const props1 = getComponent().props();
      getComponent().setProps({ careSetting: { display: 'Outpatient' } })
      const props2 = getComponent().props();
      expect(props1 !== props2).toEqual(true);
  });
});
