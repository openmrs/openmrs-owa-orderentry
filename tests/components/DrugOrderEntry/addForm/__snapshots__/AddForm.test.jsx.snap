// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for adding a new drug order should render component 1`] = `
<div>
  <div>
    <DosageTabs
      activeTabIndex={0}
      formType={[Function]}
      handleFormTabs={[Function]}
    >
      <DosageTab
        icon="icon-th-large"
        onClick={[Function]}
        tabIndex={0}
        tabName="Standard Dosage Â "
      >
        <StandardDose
          activateSaveButton={[Function]}
          fieldErrors={Object {}}
          fields={
            Object {
              "dispensingQuantity": "",
              "dispensingUnit": "",
              "dose": "",
              "dosingUnit": "",
              "drugInstructions": "",
              "duration": "",
              "durationUnit": "",
              "frequency": "",
              "reason": "",
              "route": "",
            }
          }
          handleCancel={[Function]}
          handleChange={[Function]}
          handleSubmit={[Function]}
          handleValidation={[Function]}
          options={
            Object {
              "dispensingUnit": Array [],
              "dosingUnit": Array [],
              "durationUnit": Array [],
              "frequency": Array [],
              "route": Array [],
            }
          }
        />
        <br />
      </DosageTab>
      <DosageTab
        icon="icon-edit"
        onClick={[Function]}
        tabIndex={0}
        tabName="Free Text"
      >
        <FreeText
          activateSaveButton={[Function]}
          fieldErrors={Object {}}
          fields={
            Object {
              "dispensingQuantity": "",
              "dispensingUnit": "",
              "dose": "",
              "dosingUnit": "",
              "drugInstructions": "",
              "duration": "",
              "durationUnit": "",
              "frequency": "",
              "reason": "",
              "route": "",
            }
          }
          handleCancel={[Function]}
          handleChange={[Function]}
          handleSubmit={[Function]}
          handleValidation={[Function]}
          options={
            Object {
              "dispensingUnit": Array [],
              "dosingUnit": Array [],
              "durationUnit": Array [],
              "frequency": Array [],
              "route": Array [],
            }
          }
        />
        <br />
      </DosageTab>
    </DosageTabs>
  </div>
</div>
`;
